File: YoutubeLinks.Blazor/Clients/LinkApiClient.cs
using YoutubeLinks.Shared.Abstractions;
using YoutubeLinks.Shared.Features.Links.Commands;
using YoutubeLinks.Shared.Features.Links.Queries;
using YoutubeLinks.Shared.Features.Links.Responses;
using static YoutubeLinks.Shared.Features.Links.Queries.GetAllLinks;

namespace YoutubeLinks.Blazor.Clients;

public interface ILinkApiClient
{
    Task<PagedList<LinkDto>> GetAllPaginatedLinks(GetAllPaginatedLinks.Query query);
    Task<IEnumerable<LinkInfoDto>> GetAllLinks(Query query);
    Task<LinkDto> GetLink(int id);
    Task CreateLink(CreateLink.Command command);
    Task UpdateLink(UpdateLink.Command command);
    Task SetLinkDownloadedFlag(SetLinkDownloadedFlag.Command command);
    Task DeleteLink(int id);
    Task<HttpResponseMessage> DownloadLink(DownloadLink.Command command);
    Task<HttpResponseMessage> DownloadSingleLink(DownloadSingleLink.Command command);
}

public class LinkApiClient : ILinkApiClient
{
    private const string Url = "api/links";
    private readonly IApiClient _apiClient;

    public LinkApiClient(IApiClient apiClient)
    {
        _apiClient = apiClient;
    }

    public async Task<PagedList<LinkDto>> GetAllPaginatedLinks(GetAllPaginatedLinks.Query query)
        => await _apiClient.Post<GetAllPaginatedLinks.Query, PagedList<LinkDto>>($"{Url}/allPaginated", query);

    public async Task<IEnumerable<LinkInfoDto>> GetAllLinks(Query query)
        => await _apiClient.Post<Query, IEnumerable<LinkInfoDto>>($"{Url}/all", query);

    public async Task<LinkDto> GetLink(int id)
        => await _apiClient.Get<LinkDto>($"{Url}/{id}");

    public async Task CreateLink(CreateLink.Command command)
        => await _apiClient.Post(Url, command);

    public async Task UpdateLink(UpdateLink.Command command)
        => await _apiClient.Put($"{Url}/{command.Id}", command);

    public async Task SetLinkDownloadedFlag(SetLinkDownloadedFlag.Command command)
        => await _apiClient.Put($"{Url}/{command.Id}/downloaded", command);

    public async Task DeleteLink(int id)
        => await _apiClient.Delete($"{Url}/{id}");

    public async Task<HttpResponseMessage> DownloadLink(DownloadLink.Command command)
        => await _apiClient.PostReturnHttpResponseMessage($"{Url}/download", command);

    public async Task<HttpResponseMessage> DownloadSingleLink(DownloadSingleLink.Command command)
        => await _apiClient.PostReturnHttpResponseMessage($"{Url}/downloadSingle", command);
}

File: youtube-links/src/clients/LinkApiClient.ts
import type { PagedList } from '@/shared/abstractions/PagedList';
import type { CreateLink } from '@/shared/features/links/commands/CreateLink';
import type { DownloadLink } from '@/shared/features/links/commands/DownloadLink';
import type { DownloadSingleLink } from '@/shared/features/links/commands/DownloadSingleLink';
import type { UpdateLink } from '@/shared/features/links/commands/UpdateLink';
import type { GetAllLinks } from '@/shared/features/links/queries/GetAllLinks';
import type { GetAllPaginatedLinks } from '@/shared/features/links/queries/GetAllPaginatedLinks';
import type { LinkDto } from '@/shared/features/links/responses/LinkDto';
import { apiClient, type IApiClient } from './ApiClient';
import type { AxiosResponse } from 'axios';

interface ILinkApiClient {
  getAllPaginatedLinks(query: GetAllPaginatedLinks.Query): Promise<PagedList<LinkDto>>;
  getAllLinks(query: GetAllLinks.Query): Promise<GetAllLinks.LinkInfoDto[]>;
  getLink(id: number): Promise<LinkDto>;
  createLink(command: CreateLink.Command): Promise<void>;
  updateLink(command: UpdateLink.Command): Promise<void>;
  deleteLink(id: number): Promise<void>;
  downloadLink(command: DownloadLink.Command): Promise<AxiosResponse>;
  downloadSingleLink(command: DownloadSingleLink.Command): Promise<AxiosResponse>;
}

class LinkApiClient implements ILinkApiClient {
  private readonly apiClient: IApiClient;
  private readonly url: string = 'api/links';

  constructor() {
    this.apiClient = apiClient;
  }

  async getAllPaginatedLinks(query: GetAllPaginatedLinks.Query): Promise<PagedList<LinkDto>> {
    return await this.apiClient.post<GetAllPaginatedLinks.Query, PagedList<LinkDto>>(
      `${this.url}/allPaginated`,
      query
    );
  }

  async getAllLinks(query: GetAllLinks.Query): Promise<GetAllLinks.LinkInfoDto[]> {
    return await this.apiClient.post<GetAllLinks.Query, GetAllLinks.LinkInfoDto[]>(
      `${this.url}/all`,
      query
    );
  }

  async getLink(id: number): Promise<LinkDto> {
    return await this.apiClient.get<LinkDto>(`${this.url}/${id}`);
  }

  async createLink(command: CreateLink.Command): Promise<void> {
    await this.apiClient.postWithoutResponse(this.url, command);
  }

  async updateLink(command: UpdateLink.Command): Promise<void> {
    await this.apiClient.put(`${this.url}/${command.id}`, command);
  }

  async deleteLink(id: number): Promise<void> {
    await this.apiClient.delete(`${this.url}/${id}`);
  }

  async downloadLink(command: DownloadLink.Command): Promise<AxiosResponse> {
    return await this.apiClient.postReturnAxiosResponse(`${this.url}/download`, command);
  }

  async downloadSingleLink(command: DownloadSingleLink.Command): Promise<AxiosResponse> {
    return await this.apiClient.postReturnAxiosResponse(`${this.url}/downloadSingle`, command);
  }
}

export const linkApiClient = new LinkApiClient();
